### PR TITLE
✨ feat(Source facebook-marketing): pick fields from configured json schema properties if present

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What

We have seen errors from facebook API like this:
```json
    {
      "error": {
        "code": 1,
        "message": "Please reduce the amount of data you're asking for, then retry your request"
      }
    }
```

ref:
- [issues/3009](https://github.com/airbytehq/oncall/issues/3009)
- [issues/4020](https://github.com/airbytehq/oncall/issues/4020)
- [issues/3857](https://github.com/airbytehq/oncall/issues/3857)

We want to provide users the ability to reduce the requested fields, so @artem1205 worked with [ticket 2100](https://github.com/airbytehq/airbyte-internal-issues/issues/2100) and the [PR 39522](https://github.com/airbytehq/airbyte/pull/39522) to "add json_schema from ConfiguredCatalog to Stream" and now in [ticket 8496](https://github.com/airbytehq/airbyte-internal-issues/issues/8496) we will implement the new cdk feature to determine the requested fields for facebook-marketing.

## How
When calling fields(), check first configured_json_schema; if provided, use properties keys as fields; if not, fallback in current behavior (use stream schema).

[**DEMO**](https://www.loom.com/share/38ceaa96ab1647778e723a285a1837a6?sid=2ce21cb8-a578-4792-b4b0-4ea5ea8890b0)

## Review guide

1. `airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_insight_streams.py`: add check for self.configured_json_schema
2. `airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_streams.py`: add check for self.configured_json_schema
3. `airbyte-integrations/connectors/source-facebook-marketing/pyproject.toml`: we need new cdk feature that pass configured_catalog selected fields from platform.
4. The rest is one unit test per base stream updated.

## User Impact
Users will now be able to reduce the requested fields simply checking/unchecking fields from stream configuration.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by artem1205 — APPROVED** [review]
LGTM!

nit: rename PR to align with Conventional Commits naming (see [Notion Page](https://www.notion.so/Conventional-Commits-880e2b15ff60400caeda879445f5b713))

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_streams.py [hunk 1]
```diff
@@ -55,12 +55,20 @@ def __init__(
         self.page_size = page_size if page_size is not None else 100
         self._filter_statuses = filter_statuses
         self._fields = None
+        self._saved_fields = None
```
Comment (girarda) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_streams.py:58]: nit: let's call the field `cached_selected_fields` so we're clear on what saved means.

An alternative could also be to add a `@cache` annotation to the method
```
@cache
def fields(self, **kwargs):
   ...
```

File: airbyte-integrations/connectors/source-facebook-marketing/unit_tests/integration/test_videos.py [hunk 2]
```diff
@@ -115,6 +116,25 @@ def test_given_one_page_when_read_then_return_records(self, http_mocker: HttpMoc
         output = self._read(config().with_account_ids([client_side_account_id]))
         assert len(output.records) == 1
 
+    @HttpMocker()
+    def test_request_fields_from_json_schema_in_configured_catalog(self, http_mocker: HttpMocker) -> None:
+        """
+        The purpose of this test is to check that the request fields are the same provided in json_request_schema inside configured catalog
```
Comment (girarda) [inline airbyte-integrations/connectors/source-facebook-marketing/unit_tests/integration/test_videos.py:122]: <3

**Review by girarda — APPROVED** [review]
great work @aldogonzalez8 !

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_streams.py (hunk 1 above)
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/base_streams.py:58]: I like the cache approach, will add this to base streams.

File: docs/integrations/sources/facebook-marketing.md [hunk 3]
```diff
@@ -205,6 +205,7 @@ The Facebook Marketing connector uses the `lookback_window` parameter to repeate
 
 | Version | Date       | Pull Request                                             | Subject                                                                                                                                                                                                                                                                                           |
 |:--------|:-----------|:---------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| 3.3.9 | 2024-07-08 | [41039](https://github.com/airbytehq/airbyte/pull/41039) | Pick request fields from configured json schema properties if present  |
```
[the remark this answers is not in the document]
Comment (aldogonzalez8) [reply docs/integrations/sources/facebook-marketing.md]: @girarda @artem1205 I have just updated the documentation. Do you feel comfortable with these instructions and wording?

image

[the remark this answers is not in the document]
Comment (girarda) [reply docs/integrations/sources/facebook-marketing.md]: looks good thank you!

**Pull request merged**